File: C++/4.cpp
#include <bits/stdc++.h>
using namespace std;

class SegmentTree {
    vector<int> tree;
    int n;

public:
    SegmentTree(int _n) {
        n = _n;
        tree.assign(4 * n, INT_MIN);
    }

    void build(const vector<int>& arr, int node, int start, int end) {
        if (start == end) {
            tree[node] = arr[start];
            return;
        }
        int mid = (start + end) / 2;
        build(arr, 2 * node + 1, start, mid);
        build(arr, 2 * node + 2, mid + 1, end);
        tree[node] = max(tree[2 * node + 1], tree[2 * node + 2]);
    }

    void update(int idx, int val, int node, int start, int end) {
        if (start == end) {
            tree[node] = val;
            return;
        }
        int mid = (start + end) / 2;
        if (idx <= mid) {
            update(idx, val, 2 * node + 1, start, mid);
        } else {
            update(idx, val, 2 * node + 2, mid + 1, end);
        }
        tree[node] = max(tree[2 * node + 1], tree[2 * node + 2]);
    }

    int query(int l, int r, int node, int start, int end) {
        if (r < start || end < l) return INT_MIN;
        if (l <= start && end <= r) return tree[node];
        int mid = (start + end) / 2;
        return max(query(l, r, 2 * node + 1, start, mid),
                   query(l, r, 2 * node + 2, mid + 1, end));
    }

    void build(const vector<int>& arr) {
        build(arr, 0, 0, n - 1);
    }

    void update(int idx, int val) {
        update(idx, val, 0, 0, n - 1);
    }

    int queryMax(int l, int r) {
        return query(l, r, 0, 0, n - 1);
    }
};

int main() {
    int N = 6;
    vector<int> prices = {10, 20, 15, 25, 30, 5};
    SegmentTree st(N);
    st.build(prices);

    cout << "Initial max in range [1,4]: " << st.queryMax(1, 4) << endl; // 20,15,25,30 -> 30

    st.update(2, 35);
    cout << "After update, max in range [1,4]: " << st.queryMax(1, 4) << endl; // 20,35,25,30 -> 35

    st.update(5, 40); 
    cout << "After update, max in range [0,5]: " << st.queryMax(0, 5) << endl; // 10,20,35,25,30,40 -> 40

    return 0;
}

File: Python/4.py
"""Segment Tree for Range Maximum Query with Point Updates"""

class SegmentTree:
    def __init__(self, data):
        self.n = len(data)
        self.tree = [float('-inf')] * (4 * self.n)
        self._build(data, 0, 0, self.n - 1)

    def _build(self, arr, node, start, end):
        if start == end:
            self.tree[node] = arr[start]
            return
        mid = (start + end) // 2
        self._build(arr, 2 * node + 1, start, mid)
        self._build(arr, 2 * node + 2, mid + 1, end)
        self.tree[node] = max(self.tree[2 * node + 1], self.tree[2 * node + 2])

    def update(self, idx, val, node=0, start=0, end=None):
        if end is None:
            end = self.n - 1
        if start == end:
            self.tree[node] = val
            return
        mid = (start + end) // 2
        if idx <= mid:
            self.update(idx, val, 2 * node + 1, start, mid)
        else:
            self.update(idx, val, 2 * node + 2, mid + 1, end)
        self.tree[node] = max(self.tree[2 * node + 1], self.tree[2 * node + 2])

    def query_max(self, l, r, node=0, start=0, end=None):
        if end is None:
            end = self.n - 1
        if r < start or end < l:
            return float('-inf')
        if l <= start and end <= r:
            return self.tree[node]
        mid = (start + end) // 2
        return max(self.query_max(l, r, 2 * node + 1, start, mid),
                   self.query_max(l, r, 2 * node + 2, mid + 1, end))

if __name__ == "__main__":
    prices = [10, 20, 15, 25, 30, 5]
    st = SegmentTree(prices)

    print(f"Initial max in range [1,4]: {st.query_max(1, 4)}")   # 30

    st.update(2, 35)
    print(f"After update, max in range [1,4]: {st.query_max(1, 4)}")  # 35

    st.update(5, 40)
    print(f"After update, max in range [0,5]: {st.query_max(0, 5)}")  # 40
